Migrate pcd to pointcloud tool

ivanpauno opened the pull request:
To test:

- Download https://github.com/PointCloudLibrary/data/blob/master/biwi_face_database/model.pcd.
- Run `ros2 run pcl_ros pcd_to_pointcloud --ros-args -p file_name:=$(cd . && pwd)/model.pcd`
- To check the point cloud is being sent periodically `ros2 topic echo /cloud_pcd` (or use any tool to visualize the pointcloud).

ivanpauno committed:
Migrate pcl_ros pcd_to_pointcloud tool

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

ivanpauno committed:
Remove log used for debugging

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

ivanpauno committed:
Undo unnecessary changes

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

ivanpauno committed:
Remove commented out code

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

SteveMacenski commented on pcl_ros/CMakeLists.txt:
```diff
@@ -118,8 +119,20 @@ ament_target_dependencies(pcl_ros_tf
-#add_executable(pcd_to_pointcloud tools/pcd_to_pointcloud.cpp)
-#target_link_libraries(pcd_to_pointcloud ${Boost_LIBRARIES} ${catkin_LIBRARIES} ${PCL_LIBRARIES})
+add_library(pcd_to_pointcloud_lib SHARED tools/pcd_to_pointcloud.cpp)
+# add_executable(pcd_to_pointcloud tools/pcd_to_pointcloud.cpp)
```
Does this need to be here? I don't see a `main` anymore for that to be possible. 

Though it might be nice to add a new main in a new file so that there is a standalone version available

SteveMacenski commented on pcl_ros/CMakeLists.txt:
```diff
@@ -118,8 +119,20 @@ ament_target_dependencies(pcl_ros_tf
-#add_executable(pcd_to_pointcloud tools/pcd_to_pointcloud.cpp)
-#target_link_libraries(pcd_to_pointcloud ${Boost_LIBRARIES} ${catkin_LIBRARIES} ${PCL_LIBRARIES})
+add_library(pcd_to_pointcloud_lib SHARED tools/pcd_to_pointcloud.cpp)
+# add_executable(pcd_to_pointcloud tools/pcd_to_pointcloud.cpp)
```
Does this need to be here? I don't see a `main` anymore for that to be possible. 

Though it might be nice to add a new main in a new file so that there is a standalone version available

ivanpauno replied:
I think I have already removed that line in https://github.com/ros-perception/perception_pcl/pull/359/commits/31a7fc0570bccc531a8b310f9375ae7ad4be9c0b.

SteveMacenski commented on pcl_ros/tools/pcd_to_pointcloud.cpp:
```diff
@@ -43,126 +43,89 @@
-      "Publishing data on topic %s with frame_id %s.", nh_.resolveName(
-        cloud_topic_).c_str(), tf_frame_.c_str());
-  }
+    tf_frame_ = this->declare_parameter("spin_period", tf_frame_);
```
spin period is tf frame?

ivanpauno replied:
lol, will fix that

SteveMacenski commented on pcl_ros/tools/pcd_to_pointcloud.cpp:
```diff
@@ -43,126 +43,89 @@
-      "Publishing data on topic %s with frame_id %s.", nh_.resolveName(
-        cloud_topic_).c_str(), tf_frame_.c_str());
-  }
+    tf_frame_ = this->declare_parameter("spin_period", tf_frame_);
```
spin period is tf frame?

SteveMacenski commented on pcl_ros/tools/pcd_to_pointcloud.cpp:
```diff
@@ -43,126 +43,89 @@
+    auto fields_list = pcl::getFieldsList(cloud_);
+    auto resolved_cloud_topic =
+      this->get_node_topics_interface()->resolve_topic_name(cloud_topic_);
+    RCLCPP_DEBUG(
```
Why not make this INFO and then delete the INFO below that is a partial subset of what this says?

SteveMacenski replied:
Why not make this INFO and then delete the INFO below that is a partial subset of what this says?

ivanpauno committed:
Address peer review comments

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

ivanpauno commented:
Thanks for the review @SteveMacenski !
Addressed your comments in https://github.com/ros-perception/perception_pcl/pull/359/commits/31a7fc0570bccc531a8b310f9375ae7ad4be9c0b.

jacobperron commented on pcl_ros/CMakeLists.txt:
```diff
@@ -118,8 +119,19 @@ ament_target_dependencies(pcl_ros_tf
+  sensor_msgs
+  pcl_conversions)
+rclcpp_components_register_node(pcd_to_pointcloud_lib
+  PLUGIN "pcl_ros::PCDGenerator"
```
nitpick: I would make the name of the plugin more closely resemble the name of the executable. I don't think `PCDGenerator` is an accurate description for what it does.

ivanpauno replied:
I picked the old name used in ROS 1.
If we want to update the name, `PCDPublisher` may be a good one.

@SteveMacenski what do you think?

SteveMacenski replied:
Yeah, more in line with the file name would be good

ivanpauno replied:
Renamed in https://github.com/ros-perception/perception_pcl/pull/359/commits/467ba51b03456b64f39dd8c50317965fafb8f2e2.

SteveMacenski commented on pcl_ros/CMakeLists.txt:
```diff
@@ -118,8 +119,19 @@ ament_target_dependencies(pcl_ros_tf
+add_library(pcd_to_pointcloud_lib SHARED tools/pcd_to_pointcloud.cpp)
+target_link_libraries(pcd_to_pointcloud_lib
+  ${PCL_LIBRARIES})
+target_include_directories(pcd_to_pointcloud_lib PUBLIC
```
I meant that more as a suggestion to actually add a `main` file option so that users don't have to only interact with this as a component (e.g. publish me a pointcloud for a test). I try to have any components also have an executable counterpart so that users can use either that they wish that is most sensible for their workflow.

jacobperron replied:
`rclcpp_components_register_node` also creates a standalone executable (by generating a main function). So you can invoke the program like: `ros2 run pcl_ros pcd_to_pointcloud`.

SteveMacenski replied:
Oh ok, I actually did not know that. Is that documented somewhere / new?

jacobperron replied:
It's documented [as part of the docstring](https://github.com/ros2/rclcpp/blob/03fa731d2345c41e5a4e436bf0730012ece96ce5/rclcpp_components/cmake/rclcpp_components_register_node.cmake#L15-L16). The feature was added in 2019, so not terribly new. I'm not sure if we call it out in any tutorials, but a lot of our demo packages use it.

SteveMacenski replied:
Ah feels like something that should be explicitly mentioned in composition tutorials. Here nor there for this PR, but I'd suggest maybe spending a few minutes if you can think of a place that this note would be good to expose to users. Many people's complaints about ROS 2 are these "hidden" features or got-ya's that aren't documented. Small actions accumulated over time make a big difference :-) 

Anyhow, no action here on that needed. I think the renaming is the only thing left.

jacobperron replied:
Unfortunately, I don't think we have a tutorial on writing and building node components. I'm only aware of [this tutorial](https://docs.ros.org/en/rolling/Tutorials/Composition.html) for interacting with components that already exist on the system.

SteveMacenski replied:
... there's no "how do you make a component" tutorial?

SteveMacenski replied:
I suppose https://docs.ros.org/en/rolling/Concepts/About-Composition.html?highlight=rclcpp_components_register_node#writing-a-component and https://docs.ros.org/en/rolling/Tutorials/Actions/Writing-a-Cpp-Action-Server-Client.html?highlight=rclcpp_components_register_node#writing-an-action-server-and-client-c are as good of a place as currently exists.

ivanpauno replied:
Yeah, as @jacobperron mentions, this already is creating an executable.
The [PR description](https://github.com/ros-perception/perception_pcl/pull/359#issue-1190196659) has some instructions of how to use it.

> I suppose https://docs.ros.org/en/rolling/Concepts/About-Composition.html?highlight=rclcpp_components_register_node#writing-a-component and https://docs.ros.org/en/rolling/Tutorials/Actions/Writing-a-Cpp-Action-Server-Client.html?highlight=rclcpp_components_register_node#writing-an-action-server-and-client-c are as good of a place as currently exists.

Yeah, I agree this should be explicitly mentioned in some of the tutorials.
Currently we're using it in many of the demos though, https://github.com/ros2/demos/blob/f008dbbeefd7da30c4133ad67655554d9f7f8519/demo_nodes_cpp/CMakeLists.txt#L112-L114.

ivanpauno committed:
rename node

Signed-off-by: Ivan Santiago Paunovic <ivanpauno@ekumenlabs.com>

SteveMacenski reviewed (approved):
LGTM

SteveMacenski commented:
@jacobperron you good with it?

jacobperron reviewed (approved):
LGTM! thanks!

lukicdarkoo commented:
This is super useful, thanks! Can we get the package published?
